4. LINQ/Task/Task/Task/LinqSamples.cs

```diff
@@ -0,0 +1,272 @@
+			}
+		}
+
+		[Category("Task")]
+		[Title("Task 2")]
+		[Description("Make a list of suppliers located in the same country and the same city for each client")]
+		public void Linq002()
+		{
+			var customers = dataSource.Customers
+				.Select(c => new
+				{
+					c.CustomerID,
+					c.Country,
+					c.City,
+					Suppliers = dataSource.Suppliers.Where(s => s.Country == c.Country && s.City == c.City)
+				});
+
+			foreach (var customer in customers)
+			{
+				foreach (var supplier in customer.Suppliers)
+				{
+					ObjectDumper.Write($"{customer.CustomerID} {customer.Country} {customer.City} {supplier.SupplierName} {supplier.Country} {supplier.City} ");
+				}
+			}
+		}
+
+		[Category("Task")]
+		[Title("Task 3")]
+		[Description("Find all customers who have orders that exceed the sum of X")]
+		public void Linq003()
+		{
+			decimal x = 10000;
+
+			var customers = dataSource.Customers.Where(c => c.Orders.Any(o => o.Total > x));
+
+			foreach (var customer in customers)
+			{
+				ObjectDumper.Write($"{customer.CustomerID} {customer.CompanyName} {customer.Orders.Max(o => o.Total)}");
+			}
+		}
+
+		[Category("Task")]
+		[Title("Task 4")]
+		[Description("Get a list of customers including the month of the year they became clients (consider client first order month as a required date)")]
+		public void Linq004()
+		{
+			var customers = dataSource.Customers
+				.Select(c => new
+				{
+					c.CustomerID,
+					c.CompanyName,
+					FirstOrderDate = c.Orders.OrderBy(o => o.OrderDate).Select(o => o.OrderDate).FirstOrDefault()
+				}
+				);
+
+			foreach (var customer in customers)
+			{
+				ObjectDumper.Write($"{customer.CustomerID} {customer.CompanyName} {customer.FirstOrderDate}");
+			}
+		}
+
+		[Category("Task")]
+		[Title("Task 5")]
+		[Description("Do the previous task, but get the list sorted by year, month, customer turnover (from the maximum to the minimum), and the client's name")]
+		public void Linq005()
+		{
+			var customers = dataSource.Customers
+				.Select(c => new
+				{
+					c.CustomerID,
+					c.CompanyName,
+					Turnover = c.Orders.Sum(o => o.Total),
+					FirstOrderDate = c.Orders.OrderBy(o => o.OrderDate).Select(o => o.OrderDate).FirstOrDefault()
+				}
+				)
+				.OrderBy(c => c.FirstOrderDate.Year)
+				.ThenBy(c => c.FirstOrderDate.Month)
+				.ThenByDescending(c => c.Turnover);
+
+			foreach (var customer in customers)
+			{
+				ObjectDumper.Write($"{customer.CustomerID} {customer.CompanyName} {customer.FirstOrderDate} {customer.Turnover}");
+			}
+		}
+
+		[Category("Task")]
+		[Title("Task 6")]
+		[Description("Specify all customers who have a non-digital postal code, or the region is not filled, or the operator code is not specified in the phone")]
+		public void Linq006()
+		{
+			var customers = dataSource.Customers
+				.Where(
+					c => (c.PostalCode != null && c.PostalCode.Any(p => p < '0' || p > '9'))
+					|| c.Region == null
+					|| c.Phone.First() != '('
```

Comment:
what if the phone will be null or empty?

Reply:
Fixed